### PR TITLE
Making docker local volume driver support external FS mount type

### PR DESCRIPTION
**Pull request by ghostplant**
**- What I did**
Allow docker local volume driver to mount fs using external mount command
**- How I did it**
The volume will be mounted by external mount only if mounting type starts with '^'.
**- How to verify it**
For all original mount type (tmpfs/nfs/..), it will be mounted by built-in mount and there will be no difference.
For mount type which starts with '^', it will be mounted by external mount command.
Test Step:

For orginal:
`# docker volume create --driver local --opt type=tmpfs --opt device=tmpfs --opt o=size=100m,uid=1000 --name foo`
`# docker run -it --rm -v foo:/foo ubuntu:xenial ls -la /foo`

For advanced sFTP:
`# apt-get install sshfs openssh-server`
`# ssh-keygen -t rsa -P ''`
`# mv ~/.ssh/id_rsa.pub ~/.ssh/authorized_keys`
`# ssh $(whoami)@localhost true # Make sure this step needn't input password`
`# docker volume create --driver local --opt type=^fuse.sshfs --opt device=$(whoami)@localhost:/tmp  --opt o=StrictHostKeyChecking=no --name sftp1`
`# docker run -it --rm -v sftp1:/sftp1 ubuntu:xenial ls -la /sftp1`

For docker swarm:
`# docker service create -p 8443:8443 --mount type=volume,volume-opt=type=^fuse.sshfs,volume-opt=device=$(whoami)@localhost:/tmp,volume-opt=o=workaround=all:o=reconnect:o=StrictHostKeyChecking=no,source=user1-home,target=/sftp1 ghostplant/flashback`

Making docker local volume driver support external FS mount type. such as glusterfs, cephfs, fused-based fs (sshfs/curlftpfs/..) and so on. Options for each kind of fs-type are configurable via '--volume-opt'. To mount volume with external mount command, type option is supposed to start with '^'.

Signed-off-by: CUI Wei <ghostplant@qq.com>

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
https://github.com/docker/docker/pull/28643
https://github.com/docker/docker/pull/28482
https://github.com/docker/docker/pull/28480

You do not need to open new PR when you update the code.
Please reuse the existing PR next time :+1: 
```console
$ git commit -a -s --amend
$ git push -f
```

**Comment by ghostplant** [comment]
Thank you. @AkihiroSuda 
I didn't find ways to delete those old ARs, and the "Reopen PR" was unpressable.

**Comment by AkihiroSuda** [comment]
cc @cpuguy83 @MHBauer @thaJeztah @stevvooe

**Comment by justincormack** [comment]
The `^` syntax is horrible. Maybe an option like `--mount mount=external` ?

**Comment by ghostplant** [comment]
That might cause much more code change if "mount=" flag is defined, since only --opt could be used and validOpts includes only 'type', 'o', 'device'.

Maybe `mount=external` could be added into validOpts, such as:
`docker volume create --driver local --opt type=fuse.sshfs --opt mount=external --opt device=$(whoami)@localhost:/tmp --opt o=StrictHostKeyChecking=no --name sftp1`

How about this one?

**Comment by stevvooe** [comment]
@ghostplant As I said in #28482, this design is still heavily confounds mounts and volumes. The concept of "mounts" in docker parameterize how things like binds, volumes and tmpfs are mounted. Please take some time to see how these are implemented, end to end, and you'll see the difference.

The simple difference is that `volume-opt`, as passed to the `local` driver, has nothing to do with the `--mount` flag. The volume options passed to the mount flag _do not_ parameterize the mount; they parameterize the volume if it is missing at the time the mount is performed. These options _may_ be passed through to a `mount` call, through the `local` driver, in the course of a volume creation, which, in the case of the `local` driver, are a syscall to `mount(3)`. __Understanding this distinction is critical when proposing design changes.__

It seems like this proposal would be better formed as an _extension_ to the `local` volume driver that adds an option for `helper`, that, if defined, will call out to a mount helper on the system to perform the mount. This design would manifest in the `--mount` flag as `volume-opt=helper=fuse.sshfs`, which is a result of how the data structure is designed.

**Comment by ghostplant** [comment]
So which one to be added into validOpts is better, `helper=external` or `helper=fuse.sshfs`?
If we choose the latter one, should we skip the `type` option even if a user set its value?

For example:
`--opt type=some-fs-type --opt helper=fuse.sshfs --opt device=$(whoami)@localhost:/tmp --opt o=StrictHostKeyChecking=no`

It seems `type` option is unnecessary if `helper` option is defined, is it reasonable to just ingore it since both `type` and `helper` is an option to fill fs-type.

**Comment by ghostplant** [comment]
I didn't see where method-A ( which proposes `type=^external-fstype` style ) and method-B ( which proposes `helper=external type=fuse.sshfs` style) confound with mounts and volumes according to the source codes, since all mounting cases are without any conflicts.

**Comment by ghostplant** [comment]
According to all of your advices, I think this design would satisfy most of person's feeling, that is:

Extend a `helper` flag to `--opt` options which originally include `['device', 'o', 'type']`, so all valid options should be `['device', 'o', 'type', 'helper'].`

Besides, we define `helper = [ 'built-in' (by default) | 'external' ]`, so if `helper` is not set or set to be `built-in`, local volume driver would perform completely as same as the orginal version.

If `helper` is set to be `external` manually, it would use external mount as the helper to mount more fs-type that includes not only fs types supported from kernel layer (tmpfs/sysfs/ext4/procfs/..).

Would this design be better supported?

**Comment by stevvooe** [comment]
To clarify, `--mount` and the option set you are proposing are _completely_ different subsystems. For example, `type` passed to mount selects the type of the mount, whereas `type` passed to the local volume driver sets the filesystem type. It seems like we are on the same page, since you're proposing volume options.

Can we just have the mount help get looked up based on the `type`?

Volume creation would be the following:

```
docker volume create --opt type=fuse.sshfs --opt device=$(whoami)@localhost:/tmp --opt o=StrictHostKeyChecking=no
```

Or, creation with a mount:

```
--mount src=sshfs-user-$(whoami),dst=/mnt,volume-opt=type=fuse.sshfs,volume-opt=device=$(whoami)@localhost:/tmp,volume-opt=o=StrictHostKeyChecking=no
```

(Note that if you pre-create the volume, you only need src and dst for this to work, since the volume will already be mounted as a local volume)

@cpuguy83 @justincormack Am I missing something here?

**Comment by ghostplant** [comment]
If we let local volume driver support it:
`docker volume create --opt type=fuse.sshfs --opt device=$(whoami)@localhost:/tmp --opt o=StrictHostKeyChecking=no`

then docker swarm will automatically support the following:
`docker swarm create .. --mount src=sshfs-user-$(whoami),dst=/mnt,volume-opt=type=fuse.sshfs,volume-opt=device=$(whoami)@localhost:/tmp,volume-opt=o=StrictHostKeyChecking=no`

Because swarm arguments exactly map with docker volume driver.

So deciding the pattern on how local volume options describe external mount will fix everything above.

**Comment by ghostplant** [comment]
I found supporting this will bring more interesting usage:

`cat <<EOF > /sbin/mount.hello`
`#!/bin/bash`
`zfs create $1 --mount-point=$2`
`zfs set-quota $1 size=10G`
`EOF`
`chmod a+x /sbin/mount.hello`

Then new fs-type will be supported
`docker volume create --opt type=hello`

**Comment by LK4D4** [comment]
@cpuguy83 @stevvooe what do you think about this? Should we proceed?

**Comment by stevvooe** [comment]
@LK4D4 We both had said that these need to options for the `local` volume driver. Such an approach would work without modifications to API or swarmkit.

I am not sure about the current state of this PR.

**Comment by LK4D4** [comment]
@ghostplant looks like we can close this PR and open with options to local drivers, right?

**Comment by ghostplant** [comment]
Do you think adding a new option "helper" (whose value could be either `built-in` or `external`) is acceptible?

If so, usage of local driver could be like this:

`docker volume create --opt type=fuse.sshfs --opt device=root@1.2.3.4:/tmp --opt helper=external`

and all platform that uses original local driver is not affected since they were not allowed to define `helper` option in old versions.

**Comment by ghostplant** [comment]
@LK4D4 Updated with options to local drivers

File: volume/local/local_unix.go [hunk 1]
```diff
@@ -23,17 +24,19 @@ var (
...
 type optsConfig struct {
 	MountType   string
 	MountOpts   string
 	MountDevice string
+	MountHelper string
 }
 
 func (o *optsConfig) String() string {
-	return fmt.Sprintf("type='%s' device='%s' o='%s'", o.MountType, o.MountDevice, o.MountOpts)
+	return fmt.Sprintf("type='%s' device='%s' o='%s' helper='%s'", o.MountType, o.MountDevice, o.MountOpts, o.MountHelper)
```
Comment (stevvooe) [inline volume/local/local_unix.go]: If you're trying to display quoted strings, use `%q`.

Comment (ghostplant) [inline volume/local/local_unix.go]: type='' / device='' / o=''  has same issues if values contain quote mark, do I need to fix them as well?

**Comment by ghostplant** [comment]
@cpuguy83 

**Case 1:**  .. (no `helper` defined) ..
This indicates `helper=built-in` by default, compatible with old versions

**Case 2:**  .. `helper=built-in` ..
The same as Case 1. This indicates `helper=built-in` implicitly

**Case 3:**  .. `helper=external` .. 
This indicates `helper=external` using `mount` from `PATH`

**Case 4:**  .. `helper=external:/bin/mount` .. 
This indicates `helper=external` using `mount` at location `/bin/mount`

Is the protocol like `Case 4` needed?

File: volume/local/local_unix.go (hunk 1 above)
Comment (ghostplant) [inline volume/local/local_unix.go]: Original source code output type value in this format:
`type='nfs' .. helper='external'`
If changing '%s' into '%q', the output would be like this:
`type="nfs" .. helper="external"`

If I only change `helper` output format, the output would be like this:
`type='nfs' .. helper="external"`

**Comment by justincormack** [comment]
Now we have volume drivers that are just containers, so are simple to
create and install, this seems like a regression. This assumes that the
user has a compatible version of the helper installed on the root fs docker
has access to which understands a set of options, which is a big ask. A
plugin volume driver can install its dependencies in known versions, and
can be portable. I think executing arbitrary local commands is a mistake.
We do not provide a way to install non containerised software on docker for
Mac for example, so you will never be able to use this.

So not LGTM.

**Comment by ghostplant** [comment]
@justincormack 
In most cases, `docker plugin` is a good choice for beginners.
In the following cases, docker plugin might not be a choice.
1) ARM platform (raspberrypi, etc) or hyper platform;
2) Private docker platform which doesn't allow downloading drivers from public network;
3) Options that 'docker plugin' that doesn't support, such as hook-able options like pre/post-mount;

If one meets any one of the cases above, he really needs plenty of works to do.

Local volume driver supporting standard mount seems to be just providing an alternative choice exactly solving above problems when people can't use docker plugin, just like both vip and dnsrr have their own suitable applications.

**Comment by ghostplant** [comment]
@justincormack 
`A plugin volume driver can install its dependencies in known versions`
I think this statement is not very correct. It seems docker 1.13 still doesn't support dependencies installation. Instead, it is just to execute an external mount which is compiled with Linux ELF64, bringing about:
1) dependencies problem: sshfs is not installed from external host;
2) binary compatibility problem: turn a cross-platform mount-script into an amd64 run-able script;
3) mount counter problem: each driver requires booting a new container, when it crashes (even restarted later), mount counter system will mess up;

**Comment by ghostplant** [comment]
In order to make `docker plugin` better supported by users in the future, I'd like to provide some useful suggestion below:

**1) docker plugin itself should  manage `mount counter` instead of let plugin driver to do this:**
After reading source code from not only glusterfs plugin driver but also sshfs plugin driver, I saw both of them maintain counter themselves, this is potential to break engine robust after the running plugin container crashes, see this case:
  `mount volume1 by container1 `
  `mount volume1 by container2 `
  `(.. volume driver crash ..)`
  `mount volume1 by container3 `
  `umount volume1 by container3 `
  `(.. volume driver counter mess up ..)`

**2) reservce a special name field or namespace to store binary architecture:**
   docker pull install xxx/glusterfs-amd64
   docker pull install xxx/glusterfs-arm32
   docker pull install xxx/glusterfs-i386
   ..
   this can let platform support hyper cloud.

**3) when docker daemon is pulling glusterfs driver, it should know whether the system already installed glusterfs-client, if not, try to installer it, otherwise, running the plugin driver as a container is useless.**

Before above 3 features are fixed, using local driver with `helper=external:mount` is a good choice.

**Comment by ghostplant** [comment]
I suggest changing 2 interfaces for the next version of docker volume API in order to avoid `mount counter` problem.

1) Update `Mount(r volume.MountRequest) volume.Response` to
   `Mount(r volume.MountRequest, firstMount bool) volume.Response`

2) Update `Unmount(r volume.MountRequest) volume.Response` to
   `Unmount(r volume.MountRequest, lastMount bool) volume.Response`

If the above interfaces not fixed, `docker plugin` is not good in product.

Besides, volume image for new version should have a different namespace with old version in case lots of current existing old volume images conflicting with fixed version.

**Comment by cpuguy83** [comment]
Why are we talking about mount counters here? Also not aware of any issues with counting mounts. If a plugin stores counts only in memory then it is a plugin bug.

**Comment by ghostplant** [comment]
An alternative approach is to update the Request structure to provide more information:

type volumeDriverProxyMountRequest struct {
    Name string
    ID   string
+   FirstMount bool
}

type volumeDriverProxyUnmountRequest struct {
    Name string
    ID   string
+   LastMount bool
}

What do you think of this one to solve `mount counter`? If fine, I will create another AR to focus on it.

**Comment by ghostplant** [comment]
@cpuguy83 
At least all current plugin driver have this bug. Only docker daemon knows whether a certain volume is shared or not.

**Comment by ghostplant** [comment]
I think let all volume drivers itseft to find out how many volume is shared is a big task. Persisting the counter value to disk is useless because `mount` command is not idempotent operation so it is hard to packaged into one transaction.

If docker daemon provide whether the volume is shared, things will be much easier to solve `mount counter`

File: volume/local/local_unix.go (hunk 1 above)
Comment (stevvooe) [inline volume/local/local_unix.go]: Just fix it. Does this format land anywhere? Single quoting like this won't properly escape anything. If this gets consumed, `helper` can just escape out by ending the quote and inject anything.

If this is just for display, its just cleaner to use `%q`.

Please leave the code base cleaner than you found it.

Comment (ghostplant) [inline volume/local/local_unix.go]: `%q` already output double quoting `"`, so I also removed all original single quoting `'`.

**Comment by ghostplant** [comment]
It seems local driver is still more stable at the moment.

**Comment by cpuguy83** [comment]
@justincormack It's not much different than what's currently available where we can specify any `type` and expect the kernel module to be loaded for it.
I see this as essentially the equivalent of underlay networking (macvlan, ipvlan).

@ghostplant Let's please try to keep the discussion focussed on this PR and not on several different topics.

The question is, do we allow users to specify a helper application for performing mounts with the local volume driver?

**Comment by ghostplant** [comment]
@cpuguy83 
I'll explain some security consideration.
if `helper` is using an external application, it is superficial that `helper` using external `mount` is unsafe.
But actually there are still many other possibilities for users to do unsafe things using volume options without supporting `helper`. For example:

1) docker volume create -d local --opt type=tmpfs --opt o=4094m ..: It seems that user can simply hack the host memory with this approach
2) docker volume create -d local --opt type=bind --opt device=/ --opt ..: It seems that user can totally seem the host fs-root directory, which is unimaginable.

So from the examples above, the principle is that we must not allow users to customize volume options if users are untrusted. For `helper` using external mount, the fact is the same that we should only allow trusted administrators to use it to simplify volume description.

**Comment by cpuguy83** [comment]
After discussing this with other maintainers, I'm in agreement with Justin, we don't want to rely on these distro-dependent helpers which are essentially exec-level plugins for a not well defined interface.

If, for instance, someone wants to mount a gluster volume, they can do exactly what this PR is doing but in the form of a volume plugin where the available helpers are well tested, controllable, and distributable.

Thanks for spending the time looking at this, though in the future it may be better to open an issue to discuss such changes.
Sorry we seem to be on a bit of a streak here! Let's definitely collaborate on issues so we can come up with solutions together.

Thanks again!

**Pull request closed without merging**